homeassistant/components/vivotek/camera.py

```diff
@@ -0,0 +1,257 @@
+        self._event_i0_status = None
+        self._event_0_key = DEFAULT_EVENT_0_KEY
+        self._ip = device_info.get(CONF_IP_ADDRESS)
+
+        self._get_param_url = "https://" + self._ip + DEFAULT_PATHS["get"]
+        self._set_param_url = "https://" + self._ip + DEFAULT_PATHS["set"]
+        self._still_image_url = "https://" + self._ip + DEFAULT_PATHS["still"]
+
+        username = device_info.get(CONF_USERNAME)
+        password = device_info.get(CONF_PASSWORD)
+
+        if username and password:
+            if self._authentication == HTTP_DIGEST_AUTHENTICATION:
+                self._auth = HTTPDigestAuth(username, password)
+            else:
+                self._auth = aiohttp.BasicAuth(username, password=password)
+                self._requests_auth = HTTPBasicAuth(username, password)
+        else:
+            self._auth = None
+
+        if device_info[CONF_STREAM_SOURCE]:
+            self._stream_source = (
+                "rtsp://%s:%s@%s:554/live.sdp",
+                username,
+                password,
+                device_info[CONF_STREAM_SOURCE],
+            )
+        else:
+            self._stream_source = None
+
+        self._brand = "Vivotek"
+        self._model = self.get_param("system_info_modelname").replace("'", "")
+
+        self._supported_features = SUPPORT_STREAM if self._stream_source else 0
+
+        self._last_url = None
+        self._last_image = None
+
+        self._motion_detection_enabled = self.event_enabled(self._event_0_key)
+
+    @property
+    def supported_features(self):
+        """Return supported features for this camera."""
+        return self._supported_features
+
+    @property
+    def frame_interval(self):
+        """Return the interval between frames of the mjpeg stream."""
+        return self._frame_interval
+
+    def event_enabled(self, event_key):
+        """Return true if event for the provided key is enabled."""
+        response = self.get_param(event_key)
+        return int(response.replace("'", "")) == 1
+
+    async def async_set_param(self, param, value):
+        """Set the value of the provided key."""
+        try:
+            websession = async_get_clientsession(self.hass, verify_ssl=self.verify_ssl)
+            with async_timeout.timeout(10):
+                response = await websession.post(
+                    self._set_param_url, auth=self._auth, data={param: value}
+                )
+            text = await response.text()
+            _LOGGER.info("Vivotek camera SET response text: %s", text)
+            return text.strip().split("=")[1]
+        except asyncio.TimeoutError:
+            _LOGGER.error("Timeout setting Vivotek camera parameter: %s", self._name)
+        except aiohttp.ClientError as err:
+            _LOGGER.error("Error setting Vivotek camera parameter: %s", err)
+
+    def get_param(self, param):
+        """Return the value of the provided key."""
+        try:
+            response = requests.get(
```

Comment:
Please extract the device specific interface, I/O and data parsing, to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services